haskell.*: Bootstrap & fix on powerpc64

OPNA2608 opened the pull request:
Built `ghc98` and its `ShellCheck` on hardware. Disabling the `hashable` tests on all GHC versions is only speculative, as I only saw it happening when building `ghc98.hashable`.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] powerpc64-linux (native, ELFv1)
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

sternenseemann commented:
Note that we are currently trying our best to remove as many older GHC versions as possible since keeping them around increases the maintenance burden not just for the Haskell subsystem, but also LLVM since some backends use pretty old versions. Currently, we are looking at removing everything < 9.4.

This change could possibly work as we could add the old GHCs purely for bootstrapping with the NCG backend of GHC for powerpc. However, I am _shivering_ at the amount of code and patches added in this PR and not necessarily looking forward to keeping this working.

I think the most promising solution for increasing is using the fact that we can cross compile GHC 9.4 completely. The idea would be to create “haskell bootstrap tools” archives from that for other platforms. This would also solve the issue for riscv64-linux etc. However, this is an unproven theory at the moment.

emilazy commented:
You can try `pkgsCross.ppc64-elfv1.haskell.compiler.ghc84` and see how far it gets. Since it has a native code generator, there’s a reasonable chance it’ll even produce a usable GHC that can be used to bootstrap later versions natively. (I suppose you’d probably want to arrange to give it natively‐compiled LLVM tools as a dependency, etc., to shrink the size of the closure that is cross‐compiled, but that’s not required for a proof‐of‐concept.)

(Came here via https://github.com/NixOS/nixpkgs/pull/440271 and can affirm that the idea of keeping all these older GHCs working scares me, especially if LLVM becomes involved. For POWER that shouldn’t matter, but then it doesn’t seem obvious that having this chain in‐tree is worth it for POWER over uploading a cross‐compiled GHC to tarballs.nixos.org or the like. I am sympathetic to source‐based bootstrapping concerns, but in this case I believe that nobody has ever compiled a GHC that can’t eventually be traced to some unexplained binary from the 90s, so IMO the version proliferation here is not worth tracing back to a slightly older binary (and even then, I somewhat prefer the approach used by our in‐tree minimal bootstrap of having separate stripped‐down derivations that do the minimum required to get to the next stage rather than burdening the wider package set). However, I am not a Haskell maintainer, so this is just an unsolicited opinion with no real weight.)

OPNA2608 commented:
> You can try `pkgsCross.ppc64-elfv1.haskell.compiler.ghc84` and see how far it gets.

If the Ampere machine is still around at NixCon tomorrow, then I'll see if I can give that a shot. Otherwise that'll have to wait until after NixCon, cus my laptop is prolly too puny to handle a build like that in a timely manner ^^*"*.

sternenseemann commented:
> You can try `pkgsCross.ppc64-elfv1.haskell.compiler.ghc84` and see how far it gets.

`ghc94`, not `ghc98`. Starting with GHC 9.6, the build system does no longer allow cross compiling GHC (though this may get fixed [eventually](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/11444)).

The big question for you is of course whether the result can actually be used on power, testing the build can be done by almost anyone. We were successful with riscv64-linux in the past.

> I am sympathetic to source‐based bootstrapping concerns, but in this case I believe that nobody has ever compiled a GHC that can’t eventually be traced to some unexplained binary from the 90s, so IMO the version proliferation here is not worth tracing back to a slightly older binary

The risk is also that these old binaries randomly stop working as happened with ppc64le, so a solution for getting fresh-ish binaries would be great. (Amjoseph determined that re-using other distribution's underexplained binaries is not an option since GHC is not sufficiently relocatable and patching in the /nix/store paths was not always feasible. However, this may have changed since.)

emilazy commented:
> `ghc94`, not `ghc98`.

I actually write `ghc84`, so we both typo’d :)

FWIW, @OPNA2608, it shouldn’t be meaningfully more expensive than any other given GHC build, unless you mean that we don’t have the cross toolchain for it cached, in which case I think it would be reasonable to ensure that gets built on Hydra like we do many other cross toolchains. And yes, I successfully cross‐compiled an *unregisterised* GHC 9.4 for RISC‐V when testing https://github.com/NixOS/nixpkgs/pull/440271; since PowerPC shouldn’t even involve LLVM (something I forgot when talking about injecting dependencies) and was presumably, uh, *registerised* before 9.4, I expect it should hopefully go quite smoothly. (Also, consider applying for access to the [community builders](https://nix-community.org/community-builders/) if you’d like time‐shared access to big machines to test Nixpkgs builds on? They have an Ampere, although of course one must be responsible with the shared resources.)

I guess what you’d actually want to cross‐build is a *static* `haskell.compiler.ghc94`, so that it can be fetched from a FOD without Nixpkgs dependencies? I’m not sure if GHC supports being built statically, though, especially since that would involve Musl. FODs that have Nix store dependencies are kind of awkward. I suppose you could treat it like the iSO or similar and just pack the entire closure in, and then do a bunch of splitting‐out and `patchelf(1)`ing. @sternenseemann probably has better ideas than me for how to productionize this.

OPNA2608 committed:
haskell.compiler.ghc966DebianBinary: init at 9.6.6

Yoinked the ppc64 build from Debian, because there are no bindists for this.

OPNA2608 committed:
haskellPackages.hashable: Disable tests on big-endian archs

xxhash-tests w{32,64}-examples fail on big-endian POWER. I imagine this is something endian-specific, not POWER-specific.

sternenseemann commented on pkgs/development/compilers/ghc/9.6.6-binary.nix:
```diff
@@ -0,0 +1,386 @@
+assert stdenv.targetPlatform == stdenv.hostPlatform;
+
+let
+  version = "9.6.6";
```
Does debian not provide 9.6.7?

sternenseemann commented on pkgs/development/compilers/ghc/9.6.6-binary.nix:
```diff
@@ -0,0 +1,386 @@
+    defaultLibc = {
+      powerpc64-linux = {
+        variantSuffix = "";
+        # Upstream doesn't release bindists for big-endian POWER anymore
```
Please clarify that debian does not provide a bindist tarball, but we are repackaging their package.

sternenseemann commented on pkgs/development/compilers/ghc/9.6.6-binary.nix:
```diff
@@ -0,0 +1,386 @@
+  # regular builds and GHC bootstrapping.
+  # This is "useful" for staying within hydra's output limits for at least the
+  # aarch64-linux architecture.
+  minimal ? false,
```
Please remove all code relating to `minimal`, it doesn't really make sense to carry this around as this wouldn't even be built by hydra.nixos.org, so this is just unnecessary maintenance burden.

sternenseemann commented on pkgs/development/compilers/ghc/9.6.6-binary.nix:
```diff
@@ -0,0 +1,386 @@
+    map ({ nixPackage, ... }: nixPackage) binDistUsed.archSpecificLibraries
+  );
+
+  libEnvVar = lib.optionalString stdenv.hostPlatform.isDarwin "DY" + "LD_LIBRARY_PATH";
```
This can go (darwin related)

sternenseemann commented on pkgs/development/compilers/ghc/9.6.6-binary.nix:
```diff
@@ -0,0 +1,386 @@
+    targetPackages.stdenv.cc
+    targetPackages.stdenv.cc.bintools
+    coreutils # for cat
+    (lib.getBin llvmPackages.llvm)
```
This looks wrong. POWER has NCG?!

sternenseemann replied:
loongarch64 needs LLVM, POWER shouldn't. Just remove `llvmPackages` from the dependencies altogether. If the `installCheckPhase` and bootstrapping GHC succeed without LLVM available in the derivation, then NCG works (which I would expect). GHC should already be using it by default.

You can leave the edits to `settings` since the LLVM backend is also available optionally if the user passes `-fllvm`.

sternenseemann commented on pkgs/development/compilers/ghc/9.6.6-binary.nix:
```diff
@@ -0,0 +1,386 @@
+    (lib.getBin llvmPackages.llvm)
+  ]
+  # On darwin, we need unwrapped bintools as well (for otool)
+  ++ lib.optionals (stdenv.targetPlatform.linker == "cctools") [
```
Darwin specific code can be removed.

sternenseemann commented on pkgs/development/compilers/ghc/9.6.6-binary.nix:
```diff
@@ -0,0 +1,386 @@
+  # of the bindist installer can find the libraries they expect.
+  # Cannot patchelf beforehand due to relative RPATHs that anticipate
+  # the final install location.
+  ${libEnvVar} = libPath;
```
Can be dropped (only used on darwin, I think).

sternenseemann commented on pkgs/development/compilers/ghc/9.6.6-binary.nix:
```diff
@@ -0,0 +1,386 @@
+    ar x $src
+    tar xf data.tar.xz
+    mkdir -p source
+    mv -t source/ usr var
```
Make sure to explicitly set `sourceRoot` to `source` (or pick a more descriptive name).

sternenseemann commented on pkgs/development/compilers/ghc/9.6.6-binary.nix:
```diff
@@ -0,0 +1,386 @@
+      )
+    # GHC has dtrace probes, which causes ld to try to open /usr/lib/libdtrace.dylib
+    # during linking
+    + lib.optionalString stdenv.hostPlatform.isDarwin ''
```
Darwin specific code can be removed.

sternenseemann commented on pkgs/development/compilers/ghc/9.6.6-binary.nix:
```diff
@@ -0,0 +1,386 @@
+          -exec sed -e '/^[a-z-]*library-dirs/a \    ${lib.getLib gmpUsed}/lib' -i {} \;
+    ''
+    # Similar for iconv and libffi on darwin
+    + lib.optionalString stdenv.hostPlatform.isDarwin ''
```
Darwin specific code can be removed.

sternenseemann commented on pkgs/development/compilers/ghc/9.6.6-binary.nix:
```diff
@@ -0,0 +1,386 @@
+  # find editline/gmp.
+  postFixup =
+    lib.optionalString (stdenv.hostPlatform.isLinux && !(binDistUsed.isStatic or false))
+      # Keep rpath as small as possible for patchelf#244.  All Elfs
```
Hmm, I wonder if this workaround is needed here, given that https://github.com/nixos/patchelf/issues/244 is closed.

Maybe `autoPatchelfHook` or similar would just work?!

OPNA2608 replied:
With `autoPatchelfHook` added, `binDistUsed.archSpecificLibraries` `map`ped into `buildInputs`, and the following `postFixup` code:

```sh
# autoPatchelfHook's automatic patching seemingly runs after postFixup, but we need it patched now to run one of the binaries
autoPatchelf $out/lib/ghc
package_db=("$out"/lib/ghc*/lib/package.conf.d)
"$out/bin/ghc-pkg" --package-db="$package_db" recache
```

… I get successful patching, but a segfault when trying to run the patched binary:

```
/nix/store/30hwk04jpwn966ycq655mj39yvk3vqjl-stdenv-linux/setup: line 261:  4306 Segmentation fault         "$out/bin/ghc-pkg" --package-db="$package_db" recache
```

And when moving the `autoPatchelf` stuff until the very end of the build, the final patched files seem borked:

```
Starting program: /nix/store/09skrmz3j2hbkj64vkhlj5l66471csiy-ghc-binary-9.6.6/lib/ghc/bin/ghc-9.6.6 -B/nix/store/09skrmz3j2hbkj64vkhlj5l66471csiy-ghc-binary-9.6.6/lib/ghc/lib --help

Program received signal SIGSEGV, Segmentation fault.
0x00003ffff7fdc848 in index () from /nix/store/81qg5q1d71kz6vlk04i4miczwbjkj4x7-glibc-2.40-66/lib/ld64.so.1
(gdb) bt
#0  0x00003ffff7fdc848 in index () from /nix/store/81qg5q1d71kz6vlk04i4miczwbjkj4x7-glibc-2.40-66/lib/ld64.so.1
#1  0x00003ffff7fa76c8 in ._dl_dst_count () from /nix/store/81qg5q1d71kz6vlk04i4miczwbjkj4x7-glibc-2.40-66/lib/ld64.so.1
#2  0x00003ffff7fa3408 in ._dl_map_object_deps () from /nix/store/81qg5q1d71kz6vlk04i4miczwbjkj4x7-glibc-2.40-66/lib/ld64.so.1
#3  0x00003ffff7fd40c8 in dl_main () from /nix/store/81qg5q1d71kz6vlk04i4miczwbjkj4x7-glibc-2.40-66/lib/ld64.so.1
#4  0x00003ffff7fcf830 in ._dl_sysdep_start () from /nix/store/81qg5q1d71kz6vlk04i4miczwbjkj4x7-glibc-2.40-66/lib/ld64.so.1
#5  0x00003ffff7fd1310 in ._dl_start_final () from /nix/store/81qg5q1d71kz6vlk04i4miczwbjkj4x7-glibc-2.40-66/lib/ld64.so.1
#6  0x00003ffff7fd1a60 in ._dl_start () from /nix/store/81qg5q1d71kz6vlk04i4miczwbjkj4x7-glibc-2.40-66/lib/ld64.so.1
#7  0x00003ffff7fd0410 in ._start () from /nix/store/81qg5q1d71kz6vlk04i4miczwbjkj4x7-glibc-2.40-66/lib/ld64.so.1
```

So maybe https://github.com/NixOS/patchelf/issues/244 is the wrong reference (I didn't add it, again just copied from the LoongArch64 commit), but the end result are still borked binaries. Here's the build log, with the `--debug` flag added to the automatic `patchelf` calls. Maybe someone more familiar with this than me (that is, ELF and patchelf) can see something devastatingly wrong going on there.

[ghc-binary-9.6.6.patchelf.broken.log](https://github.com/user-attachments/files/22414159/ghc-binary-9.6.6.patchelf.broken.log)

sternenseemann commented on pkgs/development/compilers/ghc/9.6.6-binary.nix:
```diff
@@ -0,0 +1,386 @@
+          fi
+        done
+      ''
+    + lib.optionalString stdenv.hostPlatform.isDarwin ''
```
Darwin specific code can be removed.

sternenseemann commented on pkgs/development/compilers/ghc/9.6.6-binary.nix:
```diff
@@ -0,0 +1,386 @@
+    # `pkgsMusl`.
+    platforms = builtins.attrNames ghcBinDists.${distSetName};
+    teams = [ lib.teams.haskell ];
+    broken = !(import ./common-have-ncg.nix { inherit lib stdenv version; });
```
This can be removed if we don't need LLVM (see above).

sternenseemann commented on pkgs/development/haskell-modules/configuration-ghc-9.10.x.nix:
```diff
@@ -114,6 +114,22 @@ in
+      #       tests/xxhash-tests.hs:21:
+      #       expected: 2768807632077661767
+      #        but got: 13521078365639231154
+      # I pretend I do not see it...
```
Have you reported this issue upstream? Please link it.

sternenseemann commented on pkgs/development/haskell-modules/configuration-ghc-9.12.x.nix:
```diff
@@ -138,6 +138,23 @@ with haskellLib;
   # https://gitlab.haskell.org/ghc/ghc/-/issues/25930
   generic-lens = dontCheck super.generic-lens;
 
+  hashable =
```
Why are these overrides all in the versioned GHC configurations? They don't seem to be version specific and should go into `configuration-common.nix`.

OPNA2608 replied:
Unfamiliarity with Haskell ecosystem stuff. Will move it there!

sternenseemann commented on pkgs/top-level/haskell-packages.nix:
```diff
@@ -82,6 +83,8 @@ in
 
       ghc963Binary = callPackage ../development/compilers/ghc/9.6.3-binary.nix { };
 
+      ghc966Binary = callPackage ../development/compilers/ghc/9.6.6-binary.nix { };
```
I think we should clarify that this is not a normal bindist and call this e.g. `ghc966DebianBinary` for clarity. This also frees us from packaging the `.deb` and bindists in the same derivation if we want to package official 9.6.6 bindists later.

OPNA2608 replied:
ACK, will do the rename once I've checked off everything else.

sternenseemann reviewed (changes requested):
Seems okay, overall. Though there is the thorny issue of preventing that this bitrots.

I'd say that we should not attempt to support theoretical configurations in the derivation and remove all dead code copied over from the other bindists, in particular code supporting darwin.

OPNA2608 commented on pkgs/development/compilers/ghc/9.6.6-binary.nix:
```diff
@@ -0,0 +1,386 @@
+        -e 's/"powerpc64-linux-gnu-ranlib")/"ranlib")/g' \
+        -e 's/"llc-18")/"llc")/g' \
+        -e 's/"opt-18")/"opt")/g' \
+        -e 's/"Use LibFFI", "YES"/"Use LibFFI", "NO"/g' \
```
```
error: Cannot build '/nix/store/3iaybqfxljbpkqb2fi5fgwb4zdmlh9s0-clock-0.8.4.drv'.
       Reason: builder failed with exit code 1.
       Output paths:
         /nix/store/c7dbi0hjv9yjsqis60hawvaxzah8yrxv-clock-0.8.4
       Last 25 log lines:
       > Using runghc version 9.6.6 found on system at:
       > /nix/store/85mdnkl1b2672vdfs8pk18i694lr4fpb-ghc-binary-9.6.6/bin/runghc-9.6.6
       > Using strip version 2.44 found on system at:
       > /nix/store/xxhdxz4h66w0g2d74f71w5f5ya27wyqq-gcc-wrapper-14.3.0/bin/strip
       > Using tar found on system at:
       > /nix/store/dv8a9ncwh3p750qd0xf7pnx6flca5cmm-gnutar-1.35/bin/tar
       > No uhc found
       > Running phase: buildPhase
       > Preprocessing library for clock-0.8.4..
       > Building library for clock-0.8.4..
       > [1 of 2] Compiling System.Clock     ( dist/build/System/Clock.hs, dist/build/System/Clock.o )
       >
       > In file included from /build/ghc228_0/ghc_2.c:4:0: error: 
       >
       > /nix/store/85mdnkl1b2672vdfs8pk18i694lr4fpb-ghc-binary-9.6.6/lib/ghc/lib/../lib/ppc64-linux-ghc-9.6.6/rts-1.0.2/include/rts/ghc_ffi.h:28:10: error:
       >      fatal error: ffi.h: No such file or directory
       >        28 | #include "ffi.h"
       >           |          ^~~~~~~
       >    |
       > 28 | #include "ffi.h"
       >    |          ^
       > compilation terminated.
       >
       > <no location info>: error:
       >     `gcc' failed in phase `C Compiler'. (Exit code: 1)
```

I'll see if I can get that working, so we don't need to disable this.

OPNA2608 replied:
I *think* I got it to work without disabling this?

2/4 `ghc98` builds since enabling this have been failing, but I'm unsure if that's related to the change or just flakiness i I hadn't seen so far:

```
| Run Ghc CompileHs Stage1: compiler/GHC/Utils/Containers/Internal/StrictPair.hs => _build/stage1/compiler/build/GHC/Utils/Containers/Internal/StrictPair.o
Command line: _build/stage0/bin/ghc -Wall -Wcompat -dynamic-too -hisuf hi -osuf o -hcsuf hc -static -hide-all-packages -no-user-package-db '-package-env -' '-package-db _build/stage1/inplace/package.conf.d' '-this-unit-id ghc-9.8.4-80a4' '-package-id array-0.5.8.0-d853' '-package-id base-4.19.2.0-d9c3' '-package-id binary-0.8.9.1-8f2d' '-package-id bytestring-0.12.1.0-3a6c' '-package-id containers-0.6.8-f58e' '-package-id deepseq-1.5.1.0-58b3' '-package-id directory-1.3.8.5-37ec' '-package-id exceptions-0.10.7-e4b4' '-package-id filepath-1.4.301.0-30c0' '-package-id ghc-boot-9.8.4-23e7' '-package-id ghc-heap-9.8.4-57ee' '-package-id ghci-9.8.4-accf' '-package-id hpc-0.7.0.0-f3da' '-package-id process-1.6.25.0-47f9' '-package-id semaphore-compat-1.0.0-defa' '-package-id stm-2.5.3.1-c50b' '-package-id template-haskell-2.21.0.0-a175' '-package-id time-1.12.2-ad5a' '-package-id transformers-0.6.1.0-4ba3' '-package-id unix-2.8.6.0-649d' -i -i/build/ghc-9.8.4-source/_build/stage1/compiler/build -i/build/ghc-9.8.4-source/_build/stage1/compiler/build/autogen -i/build/ghc-9.8.4-source/compiler -Irts/include -I_build/stage1/compiler/build -I_build/stage1/compiler/build/. -Icompiler/. -I/build/ghc-9.8.4-source/libraries/process/include -I/build/ghc-9.8.4-source/_build/stage1/libraries/process/build/include -I/nix/store/59nkc47dyqpdq170kjfcq770bd1jnm0d-libffi-3.5.1-dev/include -I/build/ghc-9.8.4-source/libraries/directory -I/build/ghc-9.8.4-source/_build/stage1/libraries/directory/build -I/build/ghc-9.8.4-source/libraries/unix/include -I/build/ghc-9.8.4-source/_build/stage1/libraries/unix/build/include -I/build/ghc-9.8.4-source/libraries/time/lib/include -I/build/ghc-9.8.4-source/_build/stage1/libraries/time/build/lib/include -I/build/ghc-9.8.4-source/libraries/containers/containers/include -I/build/ghc-9.8.4-source/_build/stage1/libraries/containers/containers/build/include -I/build/ghc-9.8.4-source/libraries/bytestring/include -I/build/ghc-9.8.4-source/_build/stage1/libraries/bytestring/build/include -I/build/ghc-9.8.4-source/libraries/base/include -I/build/ghc-9.8.4-source/_build/stage1/libraries/base/build/include -I/nix/store/cp2yrcm12hqqbxky6a8as5bycgw9xkl1-gmp-with-cxx-6.3.0-dev/include -I/build/ghc-9.8.4-source/libraries/ghc-bignum/include/ -I/build/ghc-9.8.4-source/_build/stage1/libraries/ghc-bignum/build/include/ -I/nix/store/8v0c36wgwbnlfrsqgmllgpf7gv2vcf7l-numactl-2.0.18-dev/include -I/nix/store/59nkc47dyqpdq170kjfcq770bd1jnm0d-libffi-3.5.1-dev/include -I/build/ghc-9.8.4-source/rts/include -I/build/ghc-9.8.4-source/_build/stage1/rts/build/include -optP-include -optP_build/stage1/compiler/build/autogen/cabal_macros.h -optP-DHAVE_INTERNAL_INTERPRETER -optP-DCAN_LOAD_DLL -outputdir _build/stage1/compiler/build -fdiagnostics-color=always -Wall -Wno-name-shadowing -Wnoncanonical-monad-instances -Wnoncanonical-monoid-instances -XHaskell2010 -XNoImplicitPrelude -XBangPatterns -XScopedTypeVariables -XMonoLocalBinds -XTypeOperators -no-global-package-db -package-db=/build/ghc-9.8.4-source/_build/stage1/inplace/package.conf.d -ghcversion-file=rts/include/ghcversion.h -ghcversion-file=rts/include/ghcversion.h -Wnoncanonical-monad-instances -optc-Wno-error=inline -c compiler/GHC/Utils/Containers/Internal/StrictPair.hs -o _build/stage1/compiler/build/GHC/Utils/Containers/Internal/StrictPair.o -fsplit-sections -O -H64m -O2 -haddock -haddock -Winvalid-haddock -Wno-deprecated-flags -Wcpp-undef
===> Command failed with error code: 1
ghc: panic! (the 'impossible' happened)
  GHC version 9.8.4:
        lookupGlobal
  ErrorWithoutFlag
    Failed to load interface for ‘GHC.Num.BigNat’.
    There are files missing in the ‘ghc-bignum-1.3’ package,
    try running 'ghc-pkg check'.
    Use -v to see a list of the files searched for.
  Call stack:
      CallStack (from HasCallStack):
        callStackDoc, called at compiler/GHC/Utils/Panic.hs:191:37 in ghc:GHC.Utils.Panic
        pprPanic, called at compiler/GHC/Tc/Utils/Env.hs:154:18 in ghc:GHC.Tc.Utils.Env
  CallStack (from HasCallStack):
    panic, called at compiler/GHC/Utils/Error.hs:503:29 in ghc:GHC.Utils.Error


Please report this as a GHC bug:  https://www.haskell.org/ghc/reportabug

Command failed
Build failed.
```

<sup><sub>(always love to see `the 'impossible' happened`…)</sub></sup>

I'll keep it enabled for now, and maybe do afew more rebuilds with it disabled / enabled to check if this is general flakiness or actually related…

OPNA2608 commented on pkgs/development/compilers/ghc/9.6.6-binary.nix:
```diff
@@ -0,0 +1,386 @@
+        -e 's/"llc-18")/"llc")/g' \
+        -e 's/"opt-18")/"opt")/g' \
+        -e 's/"Use LibFFI", "YES"/"Use LibFFI", "NO"/g' \
+        -e 's|"C compiler link flags", ""|"C compiler link flags", "-L${libffi.out}/lib -L${numactl.out}/lib"|g' \
```
> We make an effort to add this to the package db above

Where exactly does that happen in the derivation?

sternenseemann replied:
Ah we only do this stuff for numa, but in a way that won't work for the debian stuff probably. I've described this in my new review: https://github.com/NixOS/nixpkgs/pull/439258#pullrequestreview-3276900707

sternenseemann reviewed (commented):
Instead of adding flags to the settings file, we should properly update the package db of GHC. This means setting `include-dirs`, `library-dirs` and in some cases `library-dirs-static`:

For example, this is `lib/ghc-9.8.4/lib/package.conf.d/rts-1.0.2.conf`:

```
library-dirs:
    ${pkgroot}/../lib/x86_64-linux-ghc-9.8.4/rts-1.0.2
    /nix/store/hnzqas16kwcvaj2j5wb4izxdh692f9vn-elfutils-0.193/lib
    /nix/store/z6bn6ri10cax3lnyc4hx628hgdvaxgf1-numactl-2.0.18/lib
    /nix/store/i6a553gjc2i4r4p54qk5mdy3ybsiydwk-libffi-3.5.1/lib

library-dirs-static:
    ${pkgroot}/../lib/x86_64-linux-ghc-9.8.4/rts-1.0.2
    /nix/store/hnzqas16kwcvaj2j5wb4izxdh692f9vn-elfutils-0.193/lib
    /nix/store/z6bn6ri10cax3lnyc4hx628hgdvaxgf1-numactl-2.0.18/lib
    /nix/store/i6a553gjc2i4r4p54qk5mdy3ybsiydwk-libffi-3.5.1/lib

dynamic-library-dirs:
    ${pkgroot}/../lib/x86_64-linux-ghc-9.8.4
    /nix/store/hnzqas16kwcvaj2j5wb4izxdh692f9vn-elfutils-0.193/lib
    /nix/store/z6bn6ri10cax3lnyc4hx628hgdvaxgf1-numactl-2.0.18/lib
    /nix/store/i6a553gjc2i4r4p54qk5mdy3ybsiydwk-libffi-3.5.1/lib
```

Also

* You forgot to add a bootstrap path for GHC 9.10.3
* 9.12.* and ghcHEAD can be built by chaining source built GHCs in between, so bootstrapping using e.g. ghc910 on power.

OPNA2608 committed:
haskellPackages.inspection-testing: Mark broken on platforms where GHC lacks DWARF <-> register mappings

Unsure:

- if, and how reliably, this works on
  - RISC-V (https://gitlab.haskell.org/ghc/ghc/-/commit/1c479c01d909c5a234cc17b0b308a400c2e0a6f6)
  - LoongArch64 (https://gitlab.haskell.org/ghc/ghc/-/commit/652cba7ee71d2ebd0af912fcc218bc0f27237738)
- what the situation on unregisterised builds is like

OPNA2608 committed:
haskellPackages.vector: Disable tests when inspection-testing is borked

Fixes building up to ShellCheck on ppc64.

OPNA2608 commented on pkgs/development/compilers/ghc/9.6.6-debian-binary.nix:
```diff
@@ -0,0 +1,309 @@
+    # modifying the package-db directly
+    + ''
+      find . -name 'ghc-bignum*.conf' \
+          -exec sed -e '/^[a-z-]*library-dirs/a \    ${lib.getLib gmpUsed}/lib' -i {} \;
```
Also did this for `libffi` and `numactl`.

OPNA2608 commented:
Change requests should hopefully be addressed now. Also added myself as a maintainer of this <sub><sup>(cursed)</sup></sub> borrowed-from-Debian GHC package, and dropped some more things that are no longer relevant in the derivation (no compiling happening, no musl bindists involved).

OPNA2608 committed:
haskellPackages.crypton: Disable tests on ppc64

Upstream report about them not working on all archs. Only platform I can confirm these failures on myself is ppc64.

OPNA2608 committed:
haskellPackages.crypton-x509-validation: Disable tests on ppc64

OPNA2608 committed:
haskellPackages.tls: Disable tests on ppc64

OPNA2608 commented:
Rebased on master, resolved merge conflicts, built up to `shellcheck` on hardware again.

sternenseemann commented on pkgs/development/compilers/ghc/9.6.6-debian-binary.nix:
```diff
@@ -0,0 +1,268 @@
+    # Recache package db which needs to happen because
+    # we modify the package db
+    + ''
+      package_db=("$out"/lib/ghc*/lib/package.conf.d)
```
I think you don't need to use a bash array here since this will always succeed. This was just required for some bindists that sometimes used the hadrian and sometimes the old make dir layout depending on platform.

sternenseemann reviewed (commented):
Seems good overall, obviously, I can't really test anything. Those dontChecks are unfortunate, but there's not much we can do.

OPNA2608 commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -694,6 +696,22 @@ with haskellLib;
+        # https://gitlab.haskell.org/ghc/ghc/-/commit/1c479c01d909c5a234cc17b0b308a400c2e0a6f6
+        || (lib.versionAtLeast self.ghc.version "9.12.1" && isRiscV64)
+        # https://gitlab.haskell.org/ghc/ghc/-/commit/652cba7ee71d2ebd0af912fcc218bc0f27237738
+        || (lib.versionAtLeast self.ghc.version "9.14.1" && isLoongArch64)
```
No clue, don't have either of these to test. I could try making an unregisterised ppc64 build and see if that passes the tests? But the `ghc`s currently don't expose their `enableRegisterised` to the outside, so that wouldn't be usable here anyway.

Can also drop the version checks and leave it to the platform maintainers to adjust this as necessary.

sternenseemann replied:
I think it is probably simpler just to disable them completely or just care about power for now, riscv64 and loongaarch64 GHC are theoretical in Nixpkgs at the moment.

sternenseemann commented:
> I still need to give a static build a try, but doing bootstrapping via cross-compiled & tarball'd static GHC builds might be doable after all. Though I should prolly not be the person that figures out all the details of that ^^_"_.

I started working on this, but got a little stuck in terms of which direction to go and also don't have a ton of time at the moment, but hopefully I'll figure all of that out eventually.

sternenseemann reviewed (approved):
Just needs a formatting issue resolved, I think.

wolfgangwalther commented on pkgs/top-level/haskell-packages.nix:
I don't think we need this if else. This just uses `ghc9103` in both cases.

smaeul commented:
(Please let me know if there's a better place to discuss this.)

>> I still need to give a static build a try, but doing bootstrapping via cross-compiled & tarball'd static GHC builds might be doable after all. Though I should prolly not be the person that figures out all the details of that ^^_"_.
>
> I started working on this, but got a little stuck in terms of which direction to go and also don't have a ton of time at the moment, but hopefully I'll figure all of that out eventually.

I am also interested in getting cross bootstrapping working and merged. My system is using musl as system libc (`nixpkgs.hostPlatform = "powerpc64le-linux-musl"`), so the Debian binary package will not work easily. I was able to successfully bootstrap a native GHC from `aarch64-linux-musl` on nixos-25.05 with the following steps:

1. Cross-compile an unregisterised GHC 9.2.1 (to avoid https://gitlab.haskell.org/ghc/ghc/-/issues/21914) with the following overlay:

```
ghc921 = let
  flake = config.nix.registry.nixpkgs.flake;
  pkgFunc = import "${flake}/pkgs/development/compilers/ghc/common-make-native-bignum.nix" {
    version = "9.2.1";
    sha256 = "sha256-9EQBL5ehNtmUD3fN/wP9pI+UdeLtD+yWbE01xN9V90Y=";
  };
in (self.callPackage pkgFunc {
  bootPkgs = self.pkgsBuildBuild.haskell.packages.ghc8107Binary;
  buildTargetLlvmPackages = self.pkgsBuildTarget.llvmPackages_12;
  llvmPackages = self.llvmPackages_12;
  python3 = self.buildPackages.python311;
  inherit (self.buildPackages.python311Packages) sphinx;
  inherit (self.buildPackages.darwin) autoSignDarwinBinariesHook xattr;

  enableDocs = false;
  enableNativeBignum = true;
  enableShared = false;
  enableUnregisterised = true;
  libffi = null;
}).overrideAttrs (prev: {
  # One patch doesn't apply to GHC 9.2.1
  patches = lib.filter (patch: ! lib.hasInfix "fcompact-unwind" patch) prev.patches;

  # Required for glibc->musl cross, fixed already in nixos-25.11
  preConfigure = prev.preConfigure + ''
    export LOCALE_ARCHIVE="${self.buildPackages.glibcLocales}/lib/locale/locale-archive"
  '';
});
```

2. Create a tarball from that store path and make it available to the powerpc64le system. The binaries in this tarball only link to `musl` and `libiconv`. I didn't try to force static linking because the destination system already had the musl dynamic linker available. I could try making the bootstrap binaries fully static, but maybe this isn't necessary since musl is only a small dependency?
3. Replace GHC 9.2.8 in the bootstrap chain with the binary distribution:

```
ghc928 = self.stdenv.mkDerivation rec {
  dontBuild = true;

  installPhase = ''
    lib=$out/lib/ghc-${version}
    mkdir $out
    cp -a * $out
    patchelf \
      --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" \
      $lib/bin/*
    substituteInPlace \
      $out/bin/{{ghc,ghc-pkg,ghci,runghc}-${version},hp2ps,hpc,hsc2hs} \
      --replace-fail /nix/store/6jpk1jfwk9q34rkbrqgsm51q6qrk2y25-bash-interactive-powerpc64le-unknown-linux-musl-5.2p37 ${self.bash}
    substituteInPlace \
      $out/bin/{{ghc,ghc-pkg,ghci,runghc}-${version},hp2ps,hpc,hsc2hs} \
      $lib/package.conf.d/*.conf \
      --replace-fail /nix/store/y3lyzyix0k5z261x3f037bp4hv3wvnfv-ghc-musl-native-bignum-powerpc64le-unknown-linux-musl-9.2.1 $out
    substituteInPlace \
      $lib/package.conf.d/base-*.conf \
      $out/nix-support/propagated-target-target-deps \
      --replace-fail /nix/store/2ziimb6c5bl6nxvvdnv98mvr9xmdg4g1-musl-iconv-1.2.5 ${self.libiconv}
    substituteInPlace \
      $lib/settings \
      --replace-fail /nix/store/fr65r5hmph1ppwsfmrcy2avpm0f3nzif-gcc-wrapper-14.3.0 ${self.stdenv.cc} \
      --replace-fail /nix/store/dvcs0j71x7rrsv22p4xjn2avyjyiymdl-binutils-wrapper-2.44 ${self.stdenv.cc.bintools}
    rm $lib/package.conf.d/package.cache
    $out/bin/ghc-pkg recache --global
  '';

  meta.license = lib.licenses.bsd3;

  passthru = {
    enableShared = false;
    haskellCompilerName = "ghc-${version}";
    targetPrefix = "";
  };

  pname = "ghc-binary-cross";
  sourceRoot = ".";

  src = self.fetchurl {
    hash = "sha256-XgDmtoWXPRPT5u717oSEqMep3VjbVyWYcff2DlIAYEY=";
    url = "https://distfiles.smaeul.xyz/nix/ghc-musl-native-bignum-powerpc64le-unknown-linux-musl-${version}.tar.xz";
  };

  version = "9.2.1";
};
```

4. Allow hadrian to build with an unregisterised GHC:

```
ghc963 = compiler.ghc963.override (prev: {
  hadrian = compiler.ghc963.passthru.hadrian.overrideAttrs (prev: {
    configureFlags = prev.configureFlags ++ [ "-f-threaded" ];
  });
});
```

I did not have any issues with failing tests. I plan to try to reproduce this on nixpkgs master, wight GHC 9.4.8 as the cross-compiled bootstrap version.
